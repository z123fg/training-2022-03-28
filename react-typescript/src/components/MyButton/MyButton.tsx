import React, { FC, MouseEvent, ReactNode, useEffect, useState } from "react";
import {v4 as uuidv4} from "uuid";

type ButtonVariant = "contained" | "outlined" | "text"

type ButtonSize = "small" | "medium" | "large";

type ButtonColor = "primary" | "secondary" | "default";

interface IButtonProps {
    color?: ButtonColor;
    size?: ButtonSize;
    variant?: ButtonVariant;
    children?: ReactNode;
    disabled?:boolean;
    onClick?:()=>void
}

interface IClickPosition {
    x:number;
    y:number;
}

const MyButton: FC<IButtonProps> = ({
    color="primary",
    size="medium",
    variant="contained",
    disabled=false,
    onClick,
    children=<></>
}) => {
    const [clickPosition, setClickPosition] = useState<IClickPosition|null>(null);
    const [rippleArr, setRippleArr] = useState<ReactNode[]>([]);
    const handleClick = (e:MouseEvent) => {

        if(!disabled){
            const {offsetX,offsetY} = e.nativeEvent;
            setClickPosition({x:offsetX,y:offsetY});
            onClick?.();
        }
        
    }

    useEffect(() => {
        if(clickPosition !== null){
            const newRipple = (
                <div
                    data-testid="ripple-element"
                    key={uuidv4()}
                    style={{
                        position:"absolute",
                        left:clickPosition.x,
                        top:clickPosition.y,
                        borderRadius:"50%",
                        transform:"translate(-50%,-50%)"
                    }}
                    className={`btn-ripple btn-ripple-${variant}`}
                    onAnimationEnd={()=>{
                        console.log("animationend");
                        setRippleArr(prev=>{
                            let nextRippleArr = [...prev];
                            nextRippleArr.pop();
                            return nextRippleArr
                        })
                    }}
                >
                </div>
            )
            setRippleArr(prev=>[newRipple,...prev]);
        }
    },[clickPosition]);

    console.log("rippleArr",rippleArr)

    const composeClassName = () => {
        const colorVariantCls = `btn-${color}-${variant}`;
        const sizeCls = `btn-${size}`;
        const disabledCls = disabled?`btn-disabled`:"";
        return ["btn",colorVariantCls,sizeCls,disabledCls].join(" ");
    }
    return (
        <>
            <button className={composeClassName()} onClick={handleClick} disabled={disabled}>
                <span>{children}</span>{rippleArr}
            </button>
        </>
    )
}


export default MyButton