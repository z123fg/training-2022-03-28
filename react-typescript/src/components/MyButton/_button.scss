/* 
    color and variant are related

    contained primary
    contained secondary
    ..

    outline
    ...

    text
    ...


*/

@use "../../scss/variable";
@use "../../scss/mixin";
$color-map: (
  primary: (
    normal: variable.$primary-color,
    hover: variable.$primary-color-hover,
    hover-light: variable.$primary-color-hover-light,
  ),
  secondary: (
    normal: variable.$secondary-color,
    hover: variable.$secondary-color-hover,
    hover-light: variable.$secondary-color-hover-light,
  ),
  default: (
    normal: variable.$default-color,
    hover: variable.$default-color-hover,
    hover-light: variable.$default-color-hover-light,
  ),
);

@mixin color-variant-classes($color, $variant) {
  .btn-#{$color}-#{$variant} {
    @if $variant == "contained" {
      color: white;
      background-color: map-get(map-get($color-map, $color), "normal");
      border: none;
      box-shadow: variable.$elevation;
    } @else if $variant == "outlined" {
      background: none;
      color: map-get(map-get($color-map, $color), "normal");
      border: 1px
        solid
        rgba(map-get(map-get($color-map, $color), "normal"), 0.5);
    } @else {
      background: none;
      border: none;
      color: map-get(map-get($color-map, $color), "normal");
    }
  }
  .btn-default-#{$variant} {
    color: black;
  }

  .btn-#{$color}-#{$variant}:hover {
    @if $variant == "contained" {
      background-color: map-get(map-get($color-map, $color), "hover");
      box-shadow: none;
    } @else if $variant == "outlined" {
      background-color: rgba(
        map-get(map-get($color-map, $color), "normal"),
        0.08
      );
      border: 1px
        solid
        rgba(map-get(map-get($color-map, $color), "normal"), 0.5);
    } @else {
      background-color: rgba(
        map-get(map-get($color-map, $color), "normal"),
        0.08
      );
    }
  }

  .btn-#{$color}-#{$variant}:disabled {
    @if $variant == "contained" {
      background-color: variable.$disabled-color;
      box-shadow: none;
      color: variable.$disabled-font-color;
    } @else if $variant == "outlined" {
      background: none;
      color: variable.$disabled-font-color;
      border: 1px solid variable.$disabled-color;
    } @else {
      background: none;
      color: variable.$disabled-font-color;
    }
    pointer-events: none;
  }

  .btn-ripple-#{$variant} {
    @if $variant == "contained" {
      background-color: white;
    } @else {
      background-color: map-get(map-get($color-map, $color), "hover");
    }
  }
}

$colors: ("primary", "secondary", "default");
$variants: ("contained", "outlined", "text");

@mixin generate-color-variant-classes {
  @each $variant in $variants {
    @each $color in $colors {
      @include color-variant-classes($color, $variant);
    }
  }
}

@include generate-color-variant-classes();

.btn {
  position: relative;
  overflow: hidden;
  border-radius: 4px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  gap: 8px;
  cursor: pointer;
}

.btn-medium {
  min-width: 113px;
  height: 36px;
  span {
    @include mixin.font-medium;
    pointer-events: none;
  }
}

//.btn-medium span{

//}
.btn-small {
  min-width: 97px;
  height: 30px;
  span {
    @include mixin.font-small;
    pointer-events: none;
  }
}
.btn-large {
  min-width: 131px;
  height: 42px;
  span {
    @include mixin.font-large;
    pointer-events: none;
  }
}

.btn-ripple {
  pointer-events: none;
  animation: rippleEffect 0.8s;
}

@keyframes rippleEffect {
  0% {
    width: 0px;
    height: 0px;
    opacity: 0.1;
  }
  50% {
    opacity: 0.2;
  }
  100% {
    width: 150px;
    height: 150px;
    opacity: 0;
  }
}
