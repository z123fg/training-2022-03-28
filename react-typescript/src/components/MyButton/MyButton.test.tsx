import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import MyButton from "./MyButton";

//test suite
describe("MyButton test cases", () => {
    test("snapshot test", () => {
        const view = render(<MyButton>SUBMIT</MyButton>);
        expect(view.asFragment()).toMatchSnapshot();
    });

    test("MyButton component should mount without error", async () => {
        render(<MyButton>SUBMIT</MyButton>);
        const buttonEl = await screen.findByRole("button");
        expect(buttonEl).toBeInTheDocument();
    });

    test("MyButton should change its style based on variant type", async () => {
        const { rerender } = render(<MyButton >SUBMIT</MyButton>);
        const buttonEl = await screen.findByRole("button");
        expect(buttonEl).toHaveClass("btn-primary-contained");
        rerender(<MyButton variant="outlined">SUBMIT</MyButton>);
        expect(buttonEl).toHaveClass("btn-primary-outlined");

        rerender(<MyButton variant="text">SUBMIT</MyButton>);
        expect(buttonEl).toHaveClass("btn-primary-text");
    });

    test("MyButton should change its style based on color type", async () => {
        const { rerender } = render(<MyButton >SUBMIT</MyButton>);
        const buttonEl = await screen.findByRole("button");
        expect(buttonEl).toHaveClass("btn-primary-contained");
        rerender(<MyButton color="secondary">SUBMIT</MyButton>);
        expect(buttonEl).toHaveClass("btn-secondary-contained");

        rerender(<MyButton color="default">SUBMIT</MyButton>);
        expect(buttonEl).toHaveClass("btn-default-contained");
    });

    test("MyButton should change its style based on size type", async () => {
        const { rerender } = render(<MyButton >SUBMIT</MyButton>);
        const buttonEl = await screen.findByRole("button");
        expect(buttonEl).toHaveClass("btn-medium");
        rerender(<MyButton size="small">SUBMIT</MyButton>);
        expect(buttonEl).toHaveClass("btn-small");

        rerender(<MyButton size="large">SUBMIT</MyButton>);
        expect(buttonEl).toHaveClass("btn-large");
    });

    test("MyButton should be disabled with disabled props provided. onClick should be invoked on click if MyButton is not disabled", async () => {
        const mockClickHandler = jest.fn();
        const { rerender } = render(<MyButton onClick={mockClickHandler}>SUBMIT</MyButton>);
        const buttonEl = await screen.findByRole("button");
        fireEvent.click(buttonEl);
        expect(mockClickHandler).toBeCalledTimes(1);
        rerender(<MyButton onClick={mockClickHandler} disabled>SUBMIT</MyButton>);
        fireEvent.click(buttonEl);
        expect(mockClickHandler).toBeCalledTimes(1);
    });
    
    test("MyButton should have ripple animation on click", async () => {
        render(<MyButton>SUBMIT</MyButton>);
        const buttonEl = await screen.findByRole("button");
        let rippleEls = screen.queryAllByTestId("ripple-element");
        expect(rippleEls).toHaveLength(0)
        fireEvent.click(buttonEl);
        rippleEls = screen.queryAllByTestId("ripple-element");
        expect(rippleEls).toHaveLength(1);
        fireEvent.animationEnd(rippleEls[0]);
        rippleEls = screen.queryAllByTestId("ripple-element");
        expect(rippleEls).toHaveLength(0);
    });


})