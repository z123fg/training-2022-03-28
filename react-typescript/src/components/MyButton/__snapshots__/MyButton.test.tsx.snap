// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyButton test cases snapshot test 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary-contained btn-medium "
  >
    <span>
      SUBMIT
    </span>
  </button>
</DocumentFragment>
`;
